README.md

```diff
@@ -1,137 +1,326 @@
+```yaml
+cpp_name_space:
+  param_name: { # this is a leaf
+    type: int,
+    default_value: 3,
+    read_only: true,
+    description: "A read only  integer parameter with a default value of 3",
+    validation: {
+      integer_equal_value: [3]
+    }
+  }
+  nest1:
+    nest2:
+      param_name: { # this is a leaf
+        type: string_array
+      }
+```
 
-`Setting parameter failed: Invalid size for vector parameter admittance.damping_ratio. Expected 6 got 3`
+## Add parameter library generation to cmake project
 
-If you try to load a yaml file with missing required parameters
+To utilize this package, you must include it in a cmake project with `find_package`. Doing so will add a cmake macro called `generate_parameter_library`, which is used to generate the parameter library. In the cmake example, the following should be substituted for `# additional find_package ...`
+```cmake
+find_package(generate_parameter_library REQUIRED)
+```
 
-`ros2 run generate_parameter_library_example test_node --ros-args --params-file src/generate_parameter_library/example/config/missing_required.yaml`
+After adding the executable, which will depend on the generated code, `generate_parameter_library` is called with the following arguments.
 
+``` cmake
+generate_parameter_library(
+minimal_node_parameters # cmake target name for the parameter library
+src/minimal_node_parameters.yaml # path to input yaml file
+include/minimal_node/validators.hpp # Optional path to header containing custom validators
+)
+```
+The generated library then needs to be linked to the executable.
+```cmake
+target_link_libraries(minimal_node PRIVATE
+  rclcpp::rclcpp
+  rclcpp_lifecycle::rclcpp_lifecycle
+  minimal_node_parameters
+)
+```
+The above cmake is substituted in place of `# additional libraries ...`. Notably, the third argument of `generate_parameter_library` specifies a user provided header to include in the generated code. All validation functions defined in this header can be referred to in the YAML validation dictionary.
 
-you will get the error
+## Integrate generated struct into project source code
+Only a few source code changes are needed to integrate the generated parameter library code. First, the library must be included via `#include "minimal_node_parameters.hpp"`. Note that the include name is exactly the same as the generated library name specified for in cmake.
+```c++
+#include "minimal_node_parameters.hpp"
 ```
-terminate called after throwing an instance of 'rclcpp::exceptions::ParameterUninitializedException'
-  what():  parameter 'command_interfaces' is not initialized
-[ros2run]: Aborted
+In the initialization code, an instance of `ParamListener` and `Params` should be created. At this time, parameters will be declared and read. An exception will be thrown if any validation fails or any required parameters were not set.
+```c++
+param_listener = std::make_shared<minimal_node::ParamListener>(node);
+params_ = param_listener->get_params();
 ```
```

Comment:
I think you should put a note here about how the `Params` struct is structured and how one might read values out of it. Then after the example, you should have an explanation of what is in the generated code.